### PR TITLE
test cleanup: use ExitStack to reduce test boilerplate

### PR DESCRIPTION
**Pull request by jakevdp**
Part of the larger project to remove global state mutation from tests.

### CODE REVIEW

File: tests/pjit_test.py [hunk 1]
```diff
@@ -63,32 +62,17 @@
...
-  # Don't override user-specified device count, or other XLA flags.
-  if "xla_force_host_platform_device_count" not in flags_str:
-    os.environ["XLA_FLAGS"] = (flags_str +
-                               " --xla_force_host_platform_device_count=8")
-  # Clear any cached backends so new CPU backend will pick up the env var.
-  xla_bridge.get_backend.cache_clear()
-  global prev_spmd_lowering_flag
-  prev_spmd_lowering_flag = maps.SPMD_LOWERING.value
-  config.update('experimental_xmap_spmd_lowering', True)
+  # This will control the CPU devices. On TPU we always have 2 devices
```
Comment (yashk2810) [inline tests/pjit_test.py:69]: Remove this comment? TPU can have 8 devices too

Comment (jakevdp) [reply tests/pjit_test.py]: Done

File: jax/_src/test_util.py [hunk 2]
```diff
@@ -504,8 +504,9 @@ def device_supports_buffer_donation():
   )
 
 
+@contextmanager
 def set_host_platform_device_count(nr_devices: int):
-  """Returns a closure that undoes the operation."""
+  """Context manager to set host platform device count if not specified by user."""
```
Comment (yashk2810) [inline jax/_src/test_util.py:509]: Can we say that you cannot decorate methods with this? This is only supposed to be applied at the top level in setUpModule?

File: tests/pjit_test.py [hunk 3]
```diff
@@ -63,32 +62,15 @@
...
-  if "xla_force_host_platform_device_count" not in flags_str:
-    os.environ["XLA_FLAGS"] = (flags_str +
-                               " --xla_force_host_platform_device_count=8")
-  # Clear any cached backends so new CPU backend will pick up the env var.
-  xla_bridge.get_backend.cache_clear()
-  global prev_spmd_lowering_flag
-  prev_spmd_lowering_flag = maps.SPMD_LOWERING.value
-  config.update('experimental_xmap_spmd_lowering', True)
+  _exit_stack.enter_context(jtu.set_host_platform_device_count(8))
+  _exit_stack.enter_context(jtu.global_config_context(experimental_xmap_spmd_lowering=True))
```
Comment (yashk2810) [inline tests/pjit_test.py:70]: why is `experimental_xmap_spmd_lowering` outside of setUpModule?

Comment (jakevdp) [reply tests/pjit_test.py:70]: It's inside `setUpModule`

Comment (yashk2810) [reply tests/pjit_test.py:70]: Ohh.. github diffs are bad lol

File: jax/_src/test_util.py (hunk 2 above)
Comment (jakevdp) [reply jax/_src/test_util.py]: Done

**Review by yashk2810 — APPROVED** [review]

**Pull request merged**